File: src/promptconduit/adapters/cursor.py
"""
Cursor adapter for PromptConduit.

Translates Cursor hook events to canonical PromptConduit events.

Cursor Hook Events (Agent Hooks):
- beforeSubmitPrompt: User is about to submit a prompt
- beforeShellExecution: About to execute a shell command
- afterShellExecution: Completed executing a shell command
- beforeMCPExecution: About to execute an MCP tool
- afterMCPExecution: Completed executing an MCP tool
- beforeReadFile: About to read a file
- afterFileEdit: Completed editing a file
- afterAgentResponse: Agent finished responding
- afterAgentThought: Agent thinking output
- stop: Session/conversation stopped

Cursor Hook Events (Tab Hooks - inline completions):
- beforeTabFileRead: About to read a file for tab completion
- afterTabFileEdit: Completed a tab completion edit

See: https://cursor.com/docs/agent/hooks
"""

from typing import Dict, Any, Optional, List

from promptconduit.adapters.base import BaseAdapter
from promptconduit.schema.events import (
    CanonicalEvent,
    EventType,
    Tool,
    PromptPayload,
    ToolPayload,
    SessionPayload,
)


class CursorAdapter(BaseAdapter):
    """
    Adapter for Cursor hook events.

    Cursor uses a hook system similar to Claude Code, with JSON via stdin/stdout.
    Key differences:
    - Uses 'conversation_id' and 'generation_id' for session tracking
    - Has separate Agent and Tab hook categories
    - Includes 'cursor_version' in all events

    Example hook event (beforeSubmitPrompt):
        {
            "hook_event_name": "beforeSubmitPrompt",
            "conversation_id": "abc123",
            "generation_id": "gen456",
            "model": "gpt-4",
            "cursor_version": "0.42.0",
            "workspace_roots": ["/path/to/project"],
            "prompt": "Write a function",
            "attachments": [{"type": "file", "path": "/path/to/file.ts"}]
        }
    """

    TOOL = Tool.CURSOR

    EVENT_MAPPING = {
        # Agent hooks
        "beforeSubmitPrompt": EventType.PROMPT_SUBMIT,
        "beforeShellExecution": EventType.SHELL_PRE,
        "afterShellExecution": EventType.SHELL_POST,
        "beforeMCPExecution": EventType.TOOL_PRE,
        "afterMCPExecution": EventType.TOOL_POST,
        "beforeReadFile": EventType.FILE_READ,
        "afterFileEdit": EventType.FILE_EDIT,
        "afterAgentResponse": EventType.AGENT_RESPONSE,
        "afterAgentThought": EventType.AGENT_THOUGHT,
        "stop": EventType.SESSION_END,
        # Tab hooks (inline completions)
        "beforeTabFileRead": EventType.FILE_READ,
        "afterTabFileEdit": EventType.FILE_EDIT,
    }

    def translate_event(self, native_event: Dict[str, Any]) -> Optional[CanonicalEvent]:
        """
        Translate Cursor hook event to canonical format.

        Args:
            native_event: The raw event from Cursor's hook system

        Returns:
            CanonicalEvent if translation successful, None to skip event
        """
        native_event_name = native_event.get("hook_event_name", "")

        event_type = self.get_event_type(native_event_name)
        if not event_type:
            return None

        event = self.create_base_event(event_type, native_event, native_event_name)

        # Map Cursor's session tracking (uses conversation_id or generation_id)
        event.session_id = native_event.get("conversation_id") or native_event.get(
            "generation_id"
        )

        # Handle workspace roots (Cursor may have multiple)
        workspace_roots = native_event.get("workspace_roots", [])
        if workspace_roots and event.workspace:
            # Use the first workspace root as the repo path
            event.workspace.repo_path = workspace_roots[0]
            if not event.workspace.working_directory:
                event.workspace.working_directory = workspace_roots[0]

        # Populate event-specific payload
        if event_type == EventType.PROMPT_SUBMIT:
            event.prompt = self._translate_prompt(native_event)
        elif event_type in (EventType.SHELL_PRE, EventType.SHELL_POST):
            event.tool_event = self._translate_shell(native_event, event_type)
        elif event_type in (EventType.TOOL_PRE, EventType.TOOL_POST):
            event.tool_event = self._translate_mcp(native_event, event_type)
        elif event_type in (EventType.FILE_READ, EventType.FILE_EDIT):
            event.tool_event = self._translate_file_op(native_event, event_type)

        return event

    def _translate_prompt(self, native_event: Dict[str, Any]) -> PromptPayload:
        """
        Translate beforeSubmitPrompt to PromptPayload.

        Cursor's prompt event includes:
        - prompt: The user's input text
        - attachments: Array of file/rule attachments

        Args:
            native_event: The raw event data

        Returns:
            PromptPayload with prompt content and attachments
        """
        raw_attachments = native_event.get("attachments", [])
        attachments = None

        if raw_attachments:
            attachments = [
                {
                    "type": a.get("type"),
                    "path": a.get("path"),
                }
                for a in raw_attachments
                if isinstance(a, dict)
            ]

        return PromptPayload(
            prompt=native_event.get("prompt", ""),
            attachments=attachments if attachments else None,
        )

    def _translate_shell(
        self, native_event: Dict[str, Any], event_type: EventType
    ) -> ToolPayload:
        """
        Translate shell execution events.

        beforeShellExecution includes:
        - command: The shell command
        - cwd: Working directory

        afterShellExecution includes:
        - command: The executed command
        - output: Full terminal output
        - duration: Execution time in milliseconds

        Args:
            native_event: The raw event data
            event_type: Whether this is pre or post execution

        Returns:
            ToolPayload with shell execution details
        """
        is_post = event_type == EventType.SHELL_POST

        return ToolPayload(
            tool_name="shell",
            input={
                "command": native_event.get("command"),
                "cwd": native_event.get("cwd"),
            },
            output={"stdout": native_event.get("output")} if is_post else None,
            success=True if is_post else None,  # Cursor doesn't provide explicit success
            duration_ms=native_event.get("duration"),
        )

    def _translate_mcp(
        self, native_event: Dict[str, Any], event_type: EventType
    ) -> ToolPayload:
        """
        Translate MCP (Model Context Protocol) execution events.

        beforeMCPExecution includes:
        - tool_name: Name of the MCP tool
        - params: Tool parameters (JSON)
        - server_url or server_command: MCP server identifier

        afterMCPExecution includes:
        - tool_name: Name of the MCP tool
        - params: Input parameters
        - result: Tool output (JSON)
        - duration: Execution time in milliseconds

        Args:
            native_event: The raw event data
            event_type: Whether this is pre or post execution

        Returns:
            ToolPayload with MCP tool execution details
        """
        is_post = event_type == EventType.TOOL_POST

        return ToolPayload(
            tool_name=native_event.get("tool_name", "mcp"),
            input=native_event.get("params"),
            output=native_event.get("result") if is_post else None,
            success=True if is_post else None,
            duration_ms=native_event.get("duration"),
        )

    def _translate_file_op(
        self, native_event: Dict[str, Any], event_type: EventType
    ) -> ToolPayload:
        """
        Translate file read/edit events.

        beforeReadFile / beforeTabFileRead includes:
        - file_path: Path to the file
        - contents: File contents (for read)

        afterFileEdit / afterTabFileEdit includes:
        - file_path: Path to the file
        - edits: Array of edit objects with old_string/new_string

        Tab hooks also include range info (line numbers, columns).

        Args:
            native_event: The raw event data
            event_type: Whether this is read or edit

        Returns:
            ToolPayload with file operation details
        """
        is_edit = event_type == EventType.FILE_EDIT
        tool_name = "file_edit" if is_edit else "file_read"

        input_data: Dict[str, Any] = {
            "file_path": native_event.get("file_path"),
        }

        if not is_edit:
            # Include contents for read operations
            input_data["contents"] = native_event.get("contents")

        output_data = None
        if is_edit:
            edits = native_event.get("edits", [])
            output_data = {"edits": edits}

        return ToolPayload(
            tool_name=tool_name,
            input=input_data,
            output=output_data,
        )

File: internal/adapters/cursor.go
package adapters

import (
	"github.com/promptconduit/cli/internal/schema"
)

// CursorAdapter translates Cursor hook events to canonical format
type CursorAdapter struct {
	BaseAdapter
}

// NewCursorAdapter creates a new Cursor adapter
func NewCursorAdapter(includeRawEvent bool) *CursorAdapter {
	return &CursorAdapter{
		BaseAdapter: BaseAdapter{
			Tool:            schema.ToolCursor,
			IncludeRawEvent: includeRawEvent,
			EventMapping: map[string]schema.EventType{
				"beforeSubmitPrompt":    schema.EventPromptSubmit,
				"beforeShellExecution":  schema.EventShellPre,
				"afterShellExecution":   schema.EventShellPost,
				"beforeMCPExecution":    schema.EventToolPre,
				"afterMCPExecution":     schema.EventToolPost,
				"beforeReadFile":        schema.EventFileRead,
				"afterFileEdit":         schema.EventFileEdit,
				"afterAgentResponse":    schema.EventAgentResponse,
				"afterAgentThought":     schema.EventAgentThought,
				"stop":                  schema.EventSessionEnd,
			},
		},
	}
}

// TranslateEvent converts a Cursor native event to canonical format
func (a *CursorAdapter) TranslateEvent(nativeEvent map[string]interface{}) *schema.CanonicalEvent {
	eventName := GetString(nativeEvent, "event")
	if eventName == "" {
		return nil
	}

	eventType, ok := a.GetEventType(eventName)
	if !ok {
		return nil
	}

	event := a.CreateBaseEvent(eventType, nativeEvent, eventName)

	// Cursor uses conversation_id or generation_id for session tracking
	if convID := GetString(nativeEvent, "conversation_id"); convID != "" {
		event.SessionID = &convID
	} else if genID := GetString(nativeEvent, "generation_id"); genID != "" {
		event.SessionID = &genID
	}

	// Handle event-specific payloads
	switch eventType {
	case schema.EventPromptSubmit:
		event.Prompt = a.translatePrompt(nativeEvent)
	case schema.EventShellPre, schema.EventShellPost:
		event.ToolEvent = a.translateShell(nativeEvent, eventType)
	case schema.EventToolPre, schema.EventToolPost:
		event.ToolEvent = a.translateMCP(nativeEvent, eventType)
	case schema.EventFileRead, schema.EventFileEdit:
		event.ToolEvent = a.translateFileOp(nativeEvent, eventType)
	case schema.EventSessionEnd:
		event.Session = &schema.SessionPayload{
			Reason: GetStringPtr(nativeEvent, "reason"),
		}
	}

	return event
}

// translatePrompt extracts prompt payload from beforeSubmitPrompt events
func (a *CursorAdapter) translatePrompt(nativeEvent map[string]interface{}) *schema.PromptPayload {
	payload := &schema.PromptPayload{}

	if prompt := GetString(nativeEvent, "prompt"); prompt != "" {
		payload.Prompt = prompt
	}

	return payload
}

// translateShell extracts shell command payload
func (a *CursorAdapter) translateShell(nativeEvent map[string]interface{}, eventType schema.EventType) *schema.ToolPayload {
	payload := &schema.ToolPayload{
		ToolName: "Shell",
	}

	// Input contains command and cwd
	input := make(map[string]interface{})
	if cmd := GetString(nativeEvent, "command"); cmd != "" {
		input["command"] = cmd
	}
	if cwd := GetString(nativeEvent, "cwd"); cwd != "" {
		input["cwd"] = cwd
	}
	if len(input) > 0 {
		payload.Input = input
	}

	// For post events, include output
	if eventType == schema.EventShellPost {
		output := make(map[string]interface{})
		if out := GetString(nativeEvent, "output"); out != "" {
			output["output"] = out
		}
		if exitCode, ok := nativeEvent["exit_code"].(float64); ok {
			output["exit_code"] = int(exitCode)
			success := exitCode == 0
			payload.Success = &success
		}
		if len(output) > 0 {
			payload.Output = output
		}

		// Duration
		if durationMs, ok := nativeEvent["duration_ms"].(float64); ok {
			dur := int(durationMs)
			payload.DurationMs = &dur
		}
	}

	return payload
}

// translateMCP extracts MCP tool payload
func (a *CursorAdapter) translateMCP(nativeEvent map[string]interface{}, eventType schema.EventType) *schema.ToolPayload {
	payload := &schema.ToolPayload{}

	if toolName := GetString(nativeEvent, "tool_name"); toolName != "" {
		payload.ToolName = toolName
	}

	if params := GetMap(nativeEvent, "params"); params != nil {
		payload.Input = params
	}

	if eventType == schema.EventToolPost {
		if result := GetMap(nativeEvent, "result"); result != nil {
			payload.Output = result
		}

		if durationMs, ok := nativeEvent["duration_ms"].(float64); ok {
			dur := int(durationMs)
			payload.DurationMs = &dur
		}

		// Check for success
		if success := GetBool(nativeEvent, "success"); success != nil {
			payload.Success = success
		}
	}

	return payload
}

// translateFileOp extracts file operation payload
func (a *CursorAdapter) translateFileOp(nativeEvent map[string]interface{}, eventType schema.EventType) *schema.ToolPayload {
	var toolName string
	if eventType == schema.EventFileRead {
		toolName = "Read"
	} else {
		toolName = "Edit"
	}

	payload := &schema.ToolPayload{
		ToolName: toolName,
	}

	input := make(map[string]interface{})
	if filePath := GetString(nativeEvent, "file_path"); filePath != "" {
		input["file_path"] = filePath
	}
	if len(input) > 0 {
		payload.Input = input
	}

	// For file edits, include the changes
	if eventType == schema.EventFileEdit {
		output := make(map[string]interface{})
		if edits := GetMap(nativeEvent, "edits"); edits != nil {
			output["edits"] = edits
		}
		if len(output) > 0 {
			payload.Output = output
		}
	}

	return payload
}

// CursorEvents returns the set of known Cursor event names
var CursorEvents = map[string]bool{
	"beforeSubmitPrompt":   true,
	"beforeShellExecution": true,
	"afterShellExecution":  true,
	"beforeMCPExecution":   true,
	"afterMCPExecution":    true,
	"beforeReadFile":       true,
	"afterFileEdit":        true,
	"afterAgentResponse":   true,
	"afterAgentThought":    true,
	"stop":                 true,
}
